File: bin/mcp/src/middleware/builtin/logging.ts
/**
 * Request/response logging middleware
 */

import { Middleware, MCPRequest, MCPResponse } from "../types.js";
import { Logger } from "../../logging/logger.js";

export function createLoggingMiddleware(logger: Logger, enableRequestLogging: boolean = true, enableResponseLogging: boolean = false): Middleware {
	return {
		name: "logging",
		order: 2,
		handler: async (request: MCPRequest, next: () => Promise<MCPResponse>): Promise<MCPResponse> => {
			const start = Date.now();

			if (enableRequestLogging) {
				logger.info(`Request: ${request.method}`, {
					params: request.params,
					metadata: request.metadata,
				});
			}

			try {
				const response = await next();
				const duration = Date.now() - start;

				if (enableResponseLogging) {
					logger.info(`Response: ${request.method}`, {
						duration,
						success: !response.isError,
						metadata: response.metadata,
					});
				}

				return response;
			} catch (error) {
				const duration = Date.now() - start;
				logger.error(`Request failed: ${request.method}`, error as Error, {
					duration,
					params: request.params,
				});
				throw error;
			}
		},
	};
}

File: NeuronMCP/internal/middleware/builtin/logging.go
package builtin

import (
	"context"
	"time"

	"github.com/pgElephant/NeuronMCP/internal/logging"
	"github.com/pgElephant/NeuronMCP/internal/middleware"
)

// LoggingMiddleware logs requests and responses
type LoggingMiddleware struct {
	logger              *logging.Logger
	enableRequestLogging  bool
	enableResponseLogging bool
}

// NewLoggingMiddleware creates a new logging middleware
func NewLoggingMiddleware(logger *logging.Logger, enableRequest, enableResponse bool) *LoggingMiddleware {
	return &LoggingMiddleware{
		logger:                logger,
		enableRequestLogging:  enableRequest,
		enableResponseLogging: enableResponse,
	}
}

// Name returns the middleware name
func (m *LoggingMiddleware) Name() string {
	return "logging"
}

// Order returns the execution order
func (m *LoggingMiddleware) Order() int {
	return 2
}

// Enabled returns whether the middleware is enabled
func (m *LoggingMiddleware) Enabled() bool {
	return true
}

// Execute executes the middleware
func (m *LoggingMiddleware) Execute(ctx context.Context, req *middleware.MCPRequest, next middleware.Handler) (*middleware.MCPResponse, error) {
	start := time.Now()

	if m.enableRequestLogging {
		m.logger.Info("Request", map[string]interface{}{
			"method":   req.Method,
			"params":   req.Params,
			"metadata": req.Metadata,
		})
	}

	resp, err := next(ctx)
	duration := time.Since(start)

	if err != nil {
		m.logger.Error("Request failed", err, map[string]interface{}{
			"method":   req.Method,
			"duration": duration,
			"params":   req.Params,
		})
		return nil, err
	}

	if m.enableResponseLogging {
		m.logger.Info("Response", map[string]interface{}{
			"method":   req.Method,
			"duration": duration,
			"success":  !resp.IsError,
			"metadata": resp.Metadata,
		})
	}

	return resp, nil
}

